### PR TITLE
Update CI to Swift 6

### PR DESCRIPTION
**Pull request by finagolfin**

### CODE REVIEW

File: swift-android.patch [hunk 1]
```diff
@@ -16,20 +54,32 @@ index 9932b854cb6..ad3ac757665 100644
...
  
+diff --git a/swift-corelibs-foundation/CMakeLists.txt b/swift-corelibs-foundation/CMakeLists.txt
+index 7f290d16..95366592 100644
+--- a/swift-corelibs-foundation/CMakeLists.txt
++++ b/swift-corelibs-foundation/CMakeLists.txt
+@@ -55,6 +55,7 @@ set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
+ set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/lib)
+ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+ set(CMAKE_Swift_MODULE_DIRECTORY ${CMAKE_BINARY_DIR}/swift)
++set(CMAKE_SHARED_MODULE_CREATE_Swift_FLAGS "")
```
Comment (finagolfin) [inline swift-android.patch:65]: Sigh, I built fine locally with CMake 3.28, but the github image has 3.30 installed. I downloaded that CMake locally to reproduce the CI failure, and this patch got that local build working again, but it doesn't fix the CI! 😡

Going to leave it alone till later today: anyone else want to look into it, be my guest.

Comment (finagolfin) [reply swift-android.patch]: Never mind, this doesn't work locally either: I had used it with CMake 3.28 by mistake, which already works.

**Comment by finagolfin** [comment]
@marcprux, I've got good and bad news.

The good news is that the linux CI just went green with the 6.0 release and latest snapshots for the first time since the Foundation rewrite was merged a couple months ago! 😄

The bad news is that NIO cannot be cross-compiled from macOS, apple/swift-nio#2893, and seemingly nothing can be cross-compiled by the latest trunk 6.1 macOS toolchain. I will look into disabling those for now and at least keep the remaining macOS CI going.

If you want to look into any of those macOS problems, go right ahead. Once I get this in, I will look at merging your pull assembling the SDKs next.

**Comment by marcprux** [comment]
> The good news is that the linux CI just went green with the 6.0 release and latest snapshots for the first time since the Foundation rewrite was merged a couple months ago! 😄

Fantastic!!

> The bad news is that NIO cannot be cross-compiled from macOS, [apple/swift-nio#2893](https://github.com/apple/swift-nio/issues/2893), and seemingly nothing can be cross-compiled by the latest trunk 6.1 macOS toolchain. I will look into disabling those for now and at least keep the remaining macOS CI going.

The `trunk` error is surprising … `error: missing required module 'SwiftAndroid'`.

> If you want to look into any of those macOS problems, go right ahead. Once I get this in, I will look at merging your pull assembling the SDKs next.

I think disabling macOS NIO testing for `release` and `devel` (and disabling all macOS tests for `trunk`) makes sense to get the SDK building and CI passing, and then we can look closer at the individual failures. Let me know if you would like any assistance with that, but it should be a simple matter of turning off the individual tests based on the platform, like we do for `swift-argument-parser`.

BTW, do you intend to cease building for 5.10? That would be understandable – since I'm sure all the patches are hard enough to maintain just for Swift 6, let alone across multiple major releases – but it does close the door to supporting a hypothetical 5.10.2 or later release in the Swift 5 series.

**Comment by finagolfin** [comment]
> it should be a simple matter of turning off the individual tests based on the platform, like we do for swift-argument-parser

Yep, should be easy, stepping through that now.

> do you intend to cease building for 5.10?

Yes, once I get the latest release working, I always remove the prior release from the CI.

> it does close the door to supporting a hypothetical 5.10.2 or later release in the Swift 5 series

What would be the reason to do that? It wouldn't be hard to simply keep 5.10.1 building on the CI, I've just never needed to before. Obviously a major release is different and this is the first since I started distributing a 5.3.1 SDK four years ago, but I just don't know of any reason to.

**Pull request merged**